### ISSUE TITLE
Support direct transfer of .webm format from Youtube

### ISSUE THREAD

**Issue by metasj** [toolforge/video2commons#88]
_[youtube-dl](https://rg3.github.io/youtube-dl/download.html)_ supports requesting a .webm version of a video directly from Youtube.  When this works it is much faster than reencoding.  This would be a handy option when transferring from YT to Commons.  [I'm facing this issue now with a few hundred videos]

> youtube-dl -f "webm"  URL

**Comment by zhuyifei1999** [comment]
The current set format is "bestvideo+bestaudio", with prefer-free-formats. While it gives the best quality available, it has a few downsides, namely sometimes uploading extremely large sized videos and fail (4GiB+) (#79), and not always preferring webms when available. I'm just not smart enough to perfectly integrate with youtube-dl to find out the best option available.

Also note that the aim of this tool is not limited to YouTube. It has to support many other sites that does not offer webm, such as archive.org. Setting format to "webm" will... fail.

If you have some pointers I might be able to figure out. Or you can file a pull request if you have a solution.

**Comment by zhuyifei1999** [comment]
I'll retry this.

**Comment by veikk0** [comment]
I'm not really a programmer but I know a bit about using youtube-dl.

You don't actually have to settle for just one format; you can define multiple formats or format combinations (ie. audio+video), separate them with a slash, and if one format isn't available or is invalid for the site youtube-dl will automatically try the next one. youtube-dl also allows very specific format selection down to the actual codecs used (depending on what the extractor for the site supports, though). So instead of using bestvideo+bestaudio for everything, you could do something like this:

`youtube-dl -f bestvideo[vcodec=vp9]+bestaudio[acodec=opus]/bestvideo[ext=webm]+bestaudio[ext=webm]/bestvideo+bestaudio/best`

I don't think `vcodec` actually works for Youtube, but with this it just falls back on `bestvideo[ext=webm]+bestaudio[ext=webm]`, and then to `bestvideo+bestaudio` if a webm doesn't exist, and as a last resort to `best` which AFAIK works on all sites youtube-dl supports.

You can even restrict formats based on file size with the `filesize` field which could help with #79. Although this is a per-format selection so manual file size calculation would still be required when video and audio are not already muxed together. Anyway, less than 4 GiB `filesize` enforcement for video seems to work fine on Youtube:

```
youtube-dl --get-format -f "bestvideo[vcodec=vp9][filesize<4294967296]+bestaudio[acodec=opus]/bestvideo[ext=webm][filesize<4294967296]+bestaudio[ext=webm]/bestvideo+bestaudio/best" https://www.youtube.com/watch?v=lM02vNMRRB0

247 - 1280x720 (720p)+251 - audio only (DASH audio)
```

**Comment by zhuyifei1999** [comment]
The problem with that is that it is kinda YouTube-specific. I tried to make a generic 'format selector' but was never successful. Since we are mostly dealing with YouTube anyways I guess that's fine. Thanks :)

**Comment by zhuyifei1999** [comment]
I need to fix opus encoding (#15) and make vp9+opus default before this can work properly

**Comment by zhuyifei1999** [comment]
Thinking about this 'format selector' again, it's not really great. The format should be a soft requirement. That is, if there is a better resolution in an unsupported format than a supported format, this tool should pick the unsupported format and transcode it, instead of picking the supported format in a worse resolution. Resolution takes priority over codec. (If possible, quantization should also take priority, but that's difficult.)

As for #79, that should be a hard requirement. If the highest resolution is way too large, the tool should pick a lower resolution. File size takes priority over resolution.
